[Agent Builder] Add how-to: create alerts on trace data (#7369)
elastic/docs-content: explore-analyze/ai-features/agent-builder/create-alerts.md

---
navigation_title: "Create alerts on trace data"
description: "Create Kibana alerting rules on Agent Builder trace data to monitor token usage, costs, agent error rates, and tool failures."
applies_to:
  stack: ga 9.5+
  serverless: ga
products:
  - id: elasticsearch
  - id: kibana
  - id: observability
  - id: security
  - id: cloud-serverless
---

# Create alerts on {{agent-builder}} trace data

> {{agent-builder}} collects execution traces into a data stream in your {{es}} cluster. These traces record token usage, errors, latency, and tool calls, so you can create {{kib}} alerting rules that notify you when something needs your attention. For example, you can alert on a conversation that uses too many tokens, token costs that exceed a budget, an agent error rate that spikes, or a tool that fails repeatedly.

Reviewer: > token costs that exceed a budget
We don't store cost ($) data at the moment.

---
navigation_title: "Create alerts on trace data"
description: "Create Kibana alerting rules on Agent Builder trace data to monitor token usage, agent error rates, and tool failures."
applies_to:
  stack: ga 9.5+
  serverless: ga
products:
  - id: elasticsearch
  - id: kibana
  - id: observability
  - id: security
  - id: cloud-serverless
---

# Create alerts on {{agent-builder}} trace data

{{agent-builder}} collects execution traces into a data stream in your {{es}} cluster. These traces record token usage, errors, latency, and tool calls, so you can create {{kib}} alerting rules that notify you when something needs your attention. For example, you can alert on a conversation that uses too many tokens, total token usage that exceeds a budget, an agent error rate that spikes, or a tool that fails repeatedly.

> {{agent-builder}} has no dedicated alerting interface. You create standard {{kib}} rules against the trace data stream, so the rule types, check schedules, and connectors are the same ones you use elsewhere in {{kib}}.

Reviewer 2: ```suggestion
Because traces are stored in a regular data stream, you can use the same {{kib}} rule types, check schedules, and connectors as any other alerting workflow.
```

"{{agent-builder}} has no dedicated alerting interface." feels unnecessarily defensive, I`d drop it and rephrase slightly

## Prerequisites

Before you create a rule, make sure that:

* **Trace collection is on.** {{agent-builder}} must be writing traces to the `traces-agent_builder.otel-<space-id>` data stream in the space where you create the rule. Collection is on by default. <!-- TODO(#7322): link to Collect traces (collect-traces.md) once it is on main. -->
* **You can read the trace data.** The rule queries `traces-agent_builder.otel-*`, so you need read access to those data streams. See [Permissions](permissions.md). <!-- TODO(#7322): update to the #read-trace-data anchor once #7322 adds it. -->
* **You can use {{kib}} alerting.** You need privileges to create and manage rules, plus a connector to send notifications such as Slack, email, or PagerDuty. See [Set up alerting](/explore-analyze/alerting/alerts/alerting-setup.md).

## Create a rule

> Alert on the trace data with an [{{es}} query rule](/explore-analyze/alerting/alerts/rule-type-es-query.md) that runs an ES|QL query on a schedule and runs an action when the query returns matches. For a simple count-based threshold, you can use an [index threshold rule](/explore-analyze/alerting/alerts/rule-type-index-threshold.md) instead. The index threshold rule cannot sum the token fields, so use the {{es}} query rule with ES|QL for token-based alerts.

Reviewer 2: Consider moving the index threshold mention to a tip or note. It breaks the flow of the main instructions here.

3. For the query language, select **ES|QL**.
4. Enter your ES|QL query against the trace data stream for your space, for example `FROM traces-agent_builder.otel-<space-id>`. The query defines the condition, including the threshold. See [Example alerts](#example-alerts).
5. Set the alert grouping:

    * **Select a time field**: the field used to filter results by the rule's time window, for example `@timestamp`.
    * **Select alert group**: select **Create an alert for each row** to raise one alert per matching row, for example per conversation over the threshold. Select **Create an alert if matches are found** to raise a single alert when the query returns any rows.
6. Set the **time window** to define how far back the query searches, for example the last hour.
7. Set the **check interval** to define how often the rule runs. Keep it smaller than the time window to avoid gaps in detection.
8. Click **Test query** to confirm the query is valid. For an ES|QL query, the matching rows appear in a table.
9. Add an action, select a connector, then set the action frequency. See [Add actions](/explore-analyze/alerting/alerts/rule-type-es-query.md#_add_actions).
10. Click **Save**.

> :::{note}

Reviewer 2: this note might sit more logically inside step 7?

The `executions >= 20` guard avoids noisy alerts when an agent has run only a few times. `error_rate > 0.1` alerts when more than 10 percent of executions fail. Like conversation IDs, `attributes.gen_ai.agent.id` is a stable hash by default, so custom agents group by hash unless you turn on `agentBuilder:tracing:includeRealIds`.

This alert counts agent executions that fail outright. An error that the agent handles and still returns as a completed response does not set `status.code` to `Error`, so it is not counted here.

### A specific tool fails repeatedly

Alert when a tool records more than a set number of errors. Count `execute_tool` spans that have an error status and group by the tool's span name.

```esql
FROM traces-agent_builder.otel-default
| WHERE `span.name` LIKE "execute_tool *" AND status.code == "Error"
| STATS failures = COUNT(*) BY `span.name`
| WHERE failures > 5
```

Rule settings:

* **Alert group**: Create an alert for each row, so you get one alert per tool.
* **Time window**: the period to evaluate, for example the last hour.

Each alert identifies the tool by its span name, for example `execute_tool <toolId>`.

> :::{note}

Reviewer 2: Here we have two separate caveats crammed into one note. The anonymization point applies to any tool-based query, not just errors. Maybe it'd make more sense earlier, in the "Create a rule" section or as a note before the examples start?

The 9.5 limitation is about error coverage specifically and could be a note with a 9.5 applies_to?

6. Set the **time window** to define how far back the query searches, for example the last hour.
7. Set the **check interval** to define how often the rule runs. Keep it smaller than the time window to avoid gaps in detection.

    :::{note}
    ES|QL rules do not offer the **Exclude matches from previous run** option. If the check interval is smaller than the time window, a row that keeps matching can alert more than once. Choose the time window, check interval, and query so that a condition alerts as often as you want.
    :::
8. Click **Test query** to confirm the query is valid. For an ES|QL query, the matching rows appear in a table.
9. Add an action, select a connector, then set the action frequency. See [Add actions](/explore-analyze/alerting/alerts/rule-type-es-query.md#_add_actions).
10. Click **Save**.

After you save the rule, it appears on the **{{rules-ui}}** page, where you can confirm that it runs on schedule and check its status.

## Example alerts

Each example gives an ES|QL query and the rule settings to use with it. Adjust the fields, thresholds, and time windows to your environment.

> :::{note}

Reviewer 2: ```suggestion
:::{tip}
```

stacking admonitions is an anti-pattern, especially admonitions of the same type
